File: Library/DSU/dsu.cpp
struct dsu {
	ll n;
	v64 p, size;
	dsu() = default;
	dsu(ll n) : n(n) {
		forn(i, n + 1) {
			p.pb(i);
			size.pb(1);
		}
	}

	bool unite(ll a, ll b) {
		a = find(a);
		b = find(b);
		if (a != b) {
			if (size[a] < size[b]) swap(a, b);
			size[a] += size[b];
			p[b] = a;
			return true;
		}
		return false;
	}

	ll find(ll u) {
		if (p[u] == u) return u;
		// Path compression - Makes successive find runs in O(1);
		return p[u] = find(p[u]);
	}
};

File: Library/DSU/dsu.py
class DSU:
    def __init__(self, n):
        self.n = n
        self.p = list(range(n+1))
        self.size = [1]*(n+1)

    def find(self, u):
        if self.p[u] == u:
            return u

        # Path compression - Makes successive find runs in O(1);
        self.p[u] = self.find(self.p[u])
        return self.p[u]

    def unite(self, a, b):
        a, b = self.find(a), self.find(b)

        if a != b:
            if self.size[a] < self.size[b]:
                a, b = b, a
            self.size[a] += self.size[b]
            self.p[b] = a
            return True
        return False

    def same(self, a, b):
        return self.find(a) == self.find(b)
